redis-cluster/base/statefulset.yaml

```diff
@@ -0,0 +1,211 @@
+kind: StatefulSet
+metadata:
+  name: redis
+  labels:
+    app: redis
+  annotations:
+    "app.uw.systems/name": "Redis"
+    "app.uw.systems/description": "A key value store"
+    "app.uw.systems/repos.kubernetes-manifests": "https://github.com/utilitywarehouse/customer-support-manifests/databases/redis/cluster/base"
+    "app.uw.systems/tags.oss": "true"
+spec:
+  replicas: 6
+  selector:
+    matchLabels:
+      app: redis
+  serviceName: redis
+  updateStrategy:
+    type: RollingUpdate
+  podManagementPolicy: Parallel
+  template:
+    metadata:
+      labels:
+        app: redis
+    spec:
+      securityContext:
+        fsGroup: 1001
+      terminationGracePeriodSeconds: 30
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+            - podAffinityTerm:
+                labelSelector:
+                  matchExpressions:
+                    - key: app
+                      operator: In
+                      values:
+                        - redis
+                topologyKey: kubernetes.io/hostname
+              weight: 100
+      containers:
+        - name: redis
+          image: redis
+          imagePullPolicy: "IfNotPresent"
+          securityContext:
+            allowPrivilegeEscalation: false
+            capabilities:
+              drop:
+                - ALL
+            runAsGroup: 0
+            runAsNonRoot: true
+            runAsUser: 1001
+            seccompProfile:
+              type: RuntimeDefault
+          command:
+            - /bin/bash
+          args:
+            - -c
+            - /opt/scripts/start-node.sh
+          envFrom:
+            - configMapRef:
+                name: redis-env
+            - secretRef:
+                name: redis-password
+          ports:
+            - name: redis
+              containerPort: 6379
+            - name: redis-bus
+              containerPort: 16379
+          lifecycle:
+            preStop:
+              exec:
+                command:
+                  - /bin/bash
+                  - -c
+                  - /opt/scripts/prestop-redis.sh > /proc/1/fd/1
+          readinessProbe:
+            initialDelaySeconds: 10
+            periodSeconds: 5
+            timeoutSeconds: 1
+            failureThreshold: 5
+            exec:
+              command:
+                - sh
+                - -c
+                - /opt/scripts/ping.sh localhost $REDIS_PORT
+          resources:
+            limits:
+              cpu: 1000m
+              memory: 4Gi
+            requests:
+              cpu: 500m
+              memory: 1Gi
+          volumeMounts:
+            - name: redis-scripts
+              mountPath: /opt/scripts
+            - name: config
+              mountPath: /etc/redis/mounted/config
+            - name: data
+              mountPath: /data
+              readOnly: false
+            - name: tmp-config
+              mountPath: /etc/redis/config
+              readOnly: false
```

Comment:
why do we mount two configs, one of which is empty?
Equivalently: why do we `cp -r /etc/redis/mounted/config/. /etc/redis/config/` in the start-up script instead of just mounting the config there directly?

Reply:
So the container is running as non root and for that reason it has not got access to the directory by mounting an empty directory it does have access. 

We copy it because we modify it by adding its hostname to the config on startup.